### ISSUE TITLE
250812 Deep Learning 모델학습법_민병호

### ISSUE THREAD

**Issue by BH-Min-lab** [AIBootcamp14/wiki-rider#25]
## 📅 학습 날짜  
*2025-08-12*

## 📌 키워드 및 요약 정리
_퍼셉트론, 경사하강법, 역전파_

- _퍼셉트론 : 여러층의 신경망에서 각자의 신경이 데이터를 다음 신경으로 전달하는 구조_
- _경사하강법 : 목적 함수(예: 손실 함수)의 값을 최소화하기 위해, 함수의 기울기(gradient)를 따라 파라미터를 점진적으로 조정하는 최적화 알고리즘_
- _역전파 : 신경망 학습 과정에서 출력층의 오차를 입력층 방향으로 거꾸로 전파하면서, 각 가중치와 바이어스의 기울기(gradient)를 계산하는 알고리즘_

딥러닝의 구조

1. 학습에 필요한 데이터 
2. 주어진 데이터를 원하는 결과로 변환하는 모델
3. 모델의 결과에 대한 오차를 수치화하는 손실함수
4. 손실 함수의 값이 최소가 되도록 모델의 파라미터를 조정하는 최적화 알고리즘

* 성능 = 모델(구조) + 데이터 + 학습방법(손실함수, 알고리즘)

사람의 뉴런과 인공 뉴런

<img width="2401" height="681" alt="Image" src="https://github.com/user-attachments/assets/4f479c77-8400-4135-9707-86f14dff2951" />

인공신경망은 인간의 뉴런을 모방해 만들어짐.

퍼셉트론
퍼셉트론의 구조
여러층의 신경망에서 각자의 신경이 데이터를 다음 신경으로 전달하는 구조
 - 이전 인풋값들의 가중치를 통한 값이 임계치와 비교해 그 결과를 0과 1로만 출력값을 도출(활성화 or 비활성화).

<img width="1060" height="268" alt="Image" src="https://github.com/user-attachments/assets/50d38967-effe-48fb-8c80-d7cc1433cbdf" />

이전 아웃풋들에 가중치(weight) = w, 편향(bias) = b를 반영해 계산

퍼셉트론의 원리
여러층의 퍼셉트론의 조합으로 어떠한 논리 연산도 만들 수 있다.

다층 퍼셉트론
모델 개요

<img width="1442" height="644" alt="Image" src="https://github.com/user-attachments/assets/6635dad1-9be5-4ee2-80e3-29dbd6b7e0e7" />

Multi-Layer Perceptron(MLP) / 혹은 Fully-Connected Layers(FCs)

활성화 함수

<img width="1196" height="621" alt="Image" src="https://github.com/user-attachments/assets/e9915cfb-5b7a-4891-8ac2-2ec5e568c58d" />

이전 값들을 종합한 z값과 임계값을 비교해 활성/비활성 결정

<img width="1376" height="676" alt="Image" src="https://github.com/user-attachments/assets/5038ced5-7f18-4cc3-9197-e07d84508d2a" />

다양한 활성화 함수 존재. 그 중 시그모이드는 연산 과정 전체를 미분 가능하게 하여 역전파 기법을 사용할 수 있게 함.

뉴런 : 노드, z값과 활성화 함수를 내재
레이어/층 : 입력층, 은닉층, 출력층으로 이루어져 있음
엣지, 커넥션 : 가중치(w)를 의미

========================================================

경사하강법

미분과 기울기
변수의 움직임에 따른 함수값의 ‘변화’를 측정하는 도구

<img width="1019" height="693" alt="Image" src="https://github.com/user-attachments/assets/3d749ef5-b86d-4c48-8ed3-cff241aa7d1a" />

미분값을 통해 손실함수가 최소인 구간을 찾아갈 수 있다.

<img width="1319" height="245" alt="Image" src="https://github.com/user-attachments/assets/00d47774-ec57-4063-a625-72721df5788a" />

변수가 여러 개일 경우, 각 변수에 대해 미분한 편미분을 사용한다.

경사 하강법(GD)

<img width="673" height="460" alt="Image" src="https://github.com/user-attachments/assets/86aa4c21-6b0c-41ca-8857-071898cf8429" />

비용함수에 대한 미분값의 반대방향(마이너스방향)으로 파라미터를 업데이트 하여 기울기가 0인 부분(손실함수가 최소인 지점)을 찾는다.

<img width="907" height="69" alt="Image" src="https://github.com/user-attachments/assets/185f417f-64ad-48cc-97ff-88ea59cb70d1" />

경사하강법의 한계 및 해결방안.

<img width="1538" height="676" alt="Image" src="https://github.com/user-attachments/assets/2abf6ea2-5b52-4406-ac4a-60ce7f14f020" />

문제점 : 국소구간에서 최소값은 찾았으나 전역구간에서 최소값을 찾는데에 실패
해결책 1 : 모델 구조(혹은 데이터)를 바꾸면 함수 모양이 바뀜

<img width="1475" height="743" alt="Image" src="https://github.com/user-attachments/assets/4b954ce6-fc07-4db9-a185-0ad91053670b" />

해결책 2 : Learning Step 을 변경

확률적 경사 하강법(SGD)
전체 데이터셋 중 m개(batch size)로 분할한 후 평균 로스를 구하여 파라미터를 업데이트하는 방법

<img width="317" height="146" alt="Image" src="https://github.com/user-attachments/assets/145044a5-e13e-4fce-84c0-fd89610e917d" />

“전체 손실 함수는 그 일부인 m개의 미니배치의 평균 손실함수와 유사할 것이다”
	데이터의 일부를 가지고 그래디언트 벡터를 계산하고 파라미터를 업데이트하므로 컴퓨터의 메모리 및 연산 자원을 좀 더 효율적으로 활용할 수 있다.

========================================================

모델학습법 – 역전파

역전파의 등장
모든 가중치(weight)와 편향(bias)를 일일이 계산하는 대신 효율적으로 계산하기 위해 고안됨

연쇄 법칙

<img width="1418" height="733" alt="Image" src="https://github.com/user-attachments/assets/69dcba42-6fd9-4006-b795-57602dc5174b" />

변수가 여러 개일때 합성함수의 미분으로 나눠서 계산

역전파의 이해

<img width="1914" height="639" alt="Image" src="https://github.com/user-attachments/assets/8d52ff72-8b9b-45a8-886b-e64c36cdb965" />

각각의 매개변수를 순차적으로 미분값을 차례로 구해 곱한 값을 활용해서 다음 변수에 대한 미분값을 계산.

역전파(심화)
손실 함수의 기본 가정

<img width="514" height="235" alt="Image" src="https://github.com/user-attachments/assets/ed389313-62bd-41c9-a15f-3a104a4d0847" />
<img width="648" height="353" alt="Image" src="https://github.com/user-attachments/assets/254cd54b-d65c-40e1-9d70-4393c2ce2436" />

z : 뉴런별로 활성화함수 입력값
a : 활성화함수 출력값

<img width="2012" height="180" alt="Image" src="https://github.com/user-attachments/assets/8e14af01-e0e4-4071-b99c-9398c17acc3b" />

가정 1. 학습 데이터 샘플에 대한 신경망의 총 손실은, 각 데이터 샘플에 대한 손실의 합과 같다.

<img width="856" height="141" alt="Image" src="https://github.com/user-attachments/assets/e8c7e818-b2b0-4201-a76a-bdcaa38d2385" />

이 가정을 통해, 위와 같이 손실 함수의 편미분을 계산할 수 있다.
따라서 우리는 전체 손실이 아닌 각 데이터에 대한 손실에만 집중할 수 있다.

가정2. 각 학습 데이터 샘플에 대한 손실은 전체 손실함수에 대한 함수이다.

<img width="902" height="288" alt="Image" src="https://github.com/user-attachments/assets/e0eeef54-b2bc-414c-ab5d-20ff9ee82bc4" />

역전파의 기본 방정식

<img width="1286" height="169" alt="Image" src="https://github.com/user-attachments/assets/3f13b744-6c0c-42f1-976c-003654d79567" />

<img width="1466" height="312" alt="Image" src="https://github.com/user-attachments/assets/59272a81-9405-4fb8-adb8-33aa896e26a8" />

<img width="886" height="122" alt="Image" src="https://github.com/user-attachments/assets/9810ec88-6bc2-4ef1-9d07-0cab77970794" />

역전파의 기본 방정식1
활성화 함수의 입력값에 대한 손실함수의 미분값 = 에러

<img width="1232" height="751" alt="Image" src="https://github.com/user-attachments/assets/1b828ff9-361b-4758-8e0c-054c8b933703" />

역전파의 기본 방정식2
L+1번째 레이어의 오류값을 통해 L번째 레이어의 오류값을 계산

<img width="591" height="305" alt="Image" src="https://github.com/user-attachments/assets/61fc2d60-26c7-4d0b-b007-627c94e2f1b4" />

역전파의 기본 방정식3
코스트에 대한 L번째 bias

<img width="831" height="301" alt="Image" src="https://github.com/user-attachments/assets/f0d19100-bc25-4e11-9aed-c7edf257cd0d" />

역전파의 기본 방정식4
코스트에 대한 L번째 weight

<img width="2091" height="486" alt="Image" src="https://github.com/user-attachments/assets/7ffd27dd-735e-4747-a75c-1248a360f5b5" />

즉, 순전파는 입력 -> 예측 -> 코스트(손실함수)계산
역전파는 코스트로부터 에러를 계산하여 가중치와 바이어스 기울기를 계산하여 파라미터를 업데이트.

========================================================

손실함수
손실함수의 중요성

<img width="1623" height="707" alt="Image" src="https://github.com/user-attachments/assets/5d3090cc-33ce-40d6-a1f0-d791a5b1b6ee" />

손실함수가 줄어들도록 적절한 손실함수를 선택해야 함.

대표적인 손실 함수

<img width="498" height="187" alt="Image" src="https://github.com/user-attachments/assets/22daea34-27fc-4fdd-961f-bacd1d1dd8fa" />

평균제곱오차(Mean Squared Error, MSE)
-	초반 학습이 빠르나 이상치에 민감하다

<img width="492" height="178" alt="Image" src="https://github.com/user-attachments/assets/5d486721-1c4d-4f8e-b0bc-ef52ff34ae68" />

평균절대오차(Mean Absolute Error, MAE)
-	손실 함수의 값이 최소값에 가까워져도 미분값은 동일하기 때문에 점핑이 일어날 수 있으며, 최소값 근방에서 함수의 크기를 직접 줄여주어야 한다.

<img width="1020" height="158" alt="Image" src="https://github.com/user-attachments/assets/85d111e4-e6e6-41bd-9c7e-afddc7c55ec2" />

Huber Loss
오차가 일정 수준 이하일 때는 MSE, 그렇지 않을 때는 MAE를 사용하여 두 손실 함수의 장점을 결합한 방법

<img width="1553" height="310" alt="Image" src="https://github.com/user-attachments/assets/60eb9975-2d7f-47ae-9b68-3c523e116546" />

교차엔트로피(Cross Entropy, CE)
주어진 확률 변수 또는 사건 집합에 대한 두 확률 분포 간의 차이를 측정하는 함수

<img width="1055" height="152" alt="Image" src="https://github.com/user-attachments/assets/859519cf-6995-4864-a7bb-6f129cc53d20" />

Hinge Loss
모델이 만드는 결정 경계와 데이터 사이의 마진을 최대화하는 것을 목적으로 하는 손실 함수

손실함수의 해석
역전파 관점
-	레이어의 수가 증가할수록 활성화 함수의 미분값(0~1)이 계속해서 곱해져, 가중치에 따른 미분값이 0에 수렴하게 되는 문제

<img width="1739" height="605" alt="Image" src="https://github.com/user-attachments/assets/b135ee40-5c4a-49ad-849c-7baaad889429" />

Type1 MSE : 시그모이드함수가 곱해지므로 z값이 너무 크거나 작으면 기울기가 0에 가까워져 역전파에서 기울기 소실로 학습이 현저하게 느려질 수 있음.

<img width="987" height="269" alt="Image" src="https://github.com/user-attachments/assets/b63ffad1-74e2-4283-8e07-854b54c0dc00" />

Type2 CE : MSE에서 시그모이드함수의 미분값이 곱해지는 부분이 제거되어 기울기 소실을 일부 막을 수 있으나, 여전히 미분값이 곱해지므로 깊이가 깊어질수록 완벽하게 기울기 소실을 막을 수는 없음.

## ❓ 어려웠던 부분, 남길 말

- _수학적 해석을 있는 그대로 해석하자면 어렵지만, 블랙박스적인 성격의 딥러닝 구조를 수식을 통해 그 구조와 실행 단계를 이해할 수 있었다.
- 생성형 AI를 통해 가벼운 사례라도 수치를 만들어 실제로 계산을 해보면 동작원리와 개념이 이해 될 수 있다._

**Comment by parkjunohjjang12** [comment]
안녕하세요 민병호 님! 올려주신 내용 재미있게 잘 읽었습니다.

작성하신 글을 보니 딥러닝의 전반적인 내용을 잘 압축해 주셔서 쉽게 읽고 이해할 수 있었습니다.
  
이러한 기초 개념들을 조금 더 깊이 이해하면 이후 현업이나 연구에서 큰 도움이 될 것 같습니다. 이번 피드백은 이미 잘 알고 계시겠지만, 참고용 의견 정도로 봐주시면 좋겠습니다.  

---
## 1️⃣ 어려웠던 부분 피드백

현업이나 연구를 하다 보면 많은 분들이 깊게 신경 쓰지 않는 부분 중 하나가 **Backpropagation**입니다.  
PyTorch 등 라이브러리에서 `.backward()` 한 줄로 쉽게 계산이 가능하므로 직접 구현할 일은 거의 없지만, 이를 **직접 구현하고 개념을 명확히 이해**하면 추후 문제 해결 시 큰 도움이 됩니다. 

아래 자료들은 Backpropagation을 깊이 있게 설명하고 있으니 참고해 보시길 추천드립니다.  

특히 CS231n 자료는 저도 처음 공부할 때 큰 도움을 받았고, 아래 유튜브 강의는 제가 대학원에 입학했을 때 교수님의 수업을 직접 들으며 많은 인사이트를 얻었습니다. 
📄  [이준석 교수님 ML/DL Lec3: Neural Networks & Backpropagation](https://www.youtube.com/watch?v=O8FMi94Upxc&list=PL0E_1UqNACXAc-tb9z2069eDeIOpcO8A2&index=3
)
📄 [CS231n](https://cs231n.github.io/optimization-2/)  
📄 [Backpropagation Calculus – 3Blue1Brown](https://www.3blue1brown.com/lessons/backpropagation-calculus)  

---

## 2️⃣ 추가 학습 자료

딥러닝 개념을 조금 더 체계적으로 학습할 수 있는 자료를 공유드립니다. 물론 해당 과정에서도 좋은 수업을 들으시겠지만, 추가적으로 참고하시면 이해에 도움이 될 것 같으며 개인적으로 Theme 2까지 학습하시면 기본기를 탄탄히 다질 수 있고, Backpropagation 코드 구현 과제도 있어 실습까지 가능합니다.
📄 [NYU Deep Learning](https://atcold.github.io/NYU-DLSP21/)

---

## 3️⃣ 학습 팁

- Backpropagation을 직접 구현해보면서 수식과 연산 흐름을 눈으로 확인하면 이해도가 크게 올라갑니다.  
- 단순히 라이브러리 사용법을 익히는 것보다, **수식 → 코드 → 결과 해석**의 3단계 학습을 권장드립니다.  
- 시간이 된다면 위 자료의 과제를 실제로 구현해 보시길 추천드립니다.

---

감사합니다~! :)
